compaction.go

```diff
@@ -100,10 +106,433 @@ func (t *TableBlock) compact() error {
+//
+//  2. As more data is added to a granule compacted in point 1, the new data
+//     will again be stored in a suboptimal multi-part format with multiple
+//     variable length row groups. The same compaction as point 1 is performed
+//     but only on the new data (i.e. the already compacted part is left
+//     untouched). If the new part brings the granule size below the threshold,
+//     the compaction is complete. If not, a cheap split is performed, which
+//     simply means that the new part will be reassigned to a new granule.
+//
+//  3. Sometimes, an expensive split must be performed. This happens when a
+//     granule is above the threshold size and contains only one part. This
+//     requires reading the entire part into memory and rewriting the row
+//     groups into two new granules.
+//
+// Some things to be aware of:
+//   - The current compaction strategy assumes there is no benefit to having
+//     muliple parts over a single part as long as the row groups are of equal
+//     length. If this is the case, a future enhancement could be to limit the
+//     number of row groups per part so that fewer expensive splits need to be
+//     performed.
+//   - If the above is not true, then we should always aim to have the least
+//     amount of parts per granule, and should introduce a second compaction
+//     level to parts.
+//   - Currently, our compaction strategy is based on a sweep on a timed
+//     interval. This is OK for now, but ideally we would track metrics that
+//     are proxies for memory size/scan speed that would trigger compactions.
+//     For example, the ratio of uncompacted bytes to total bytes. Or performing
+//     split operations based on load.
+//
+// A note on ordering guarantees:
+// FrostDB guarantees ordering amongst granules. This means that if there are
+// two granules in the index, all rows in granule A are less than granule B.
+// However, within a granule, rows are only ordered at a part level since parts
+// are created when a client appends rows and clients are not required to
+// append in order. The current compaction strategy ensures that ordering
+// guarantees between granules are upheld. Additionally, level1 parts will never
+// overlap.
+func (t *TableBlock) compactGranule(granule *Granule) error {
+	// Use the latest watermark as the tx id.
+	tx := t.table.db.tx.Load()
+
+	level0Parts, level1Parts, partsWithUncompletedTxns, sizeBefore, err := collectPartsForCompaction(
+		// Start compaction by adding a sentinel node to its parts list.
+		tx, granule.parts.Sentinel(Compacting),
+	)
+	if err != nil {
+		return err
+	}
+
+	if len(level0Parts) == 0 && len(level1Parts) == 0 {
+		// Edge case in which only parts with uncompleted txns are found and
+		// they are already bigger than the max size. A future GC cycle will
+		// take care of this, so abort the compaction and return.
+		t.abortCompaction(granule)
```

Comment:
`nit` prefer aborting in a single place than the potential to abort in multiple places. Can we instead return an error here and have the caller abort like we do for the rest of the error cases?

Reply:
It's a good point. I did think about returning the error here but it would be logged as a warning by the higher level compactor pool and I wasn't keen on putting error code inspection in there. I ended up considering this as a successful "edge case" and aborting here. I'm not 100% sure this is the right path so I'd be keen on getting your perspective.

Reply:
I ended up returning a boolean to accompany the error.